chemper/graphs/cluster_graph.py

```diff
@@ -262,6 +297,128 @@ def _add_first_smirks_atoms(self, mol, smirks_atoms):
                                          neighbor_storage,
                                          bond=bond_storage)
 
+        for smirks_index, atom_index in smirks_atoms.items():
+            atom = mol.get_atom_by_index(atom_index)
+            storage = self.atom_by_smirks_index[smirks_index]
+            self._add_layers(mol, atom, storage, self.layers, atom_dict)
+
+    def _add_layers(self, mol, atom, storage, layers, idx_dict):
```

Comment:
This part is also probably involved enough you should have comments that briefly explain the algorithm to make the code easier to follow.